dds_cli/user.py

```diff
@@ -99,12 +99,48 @@ def __authenticate_user(self):
                 message="Non-empty password needed to be able to authenticate."
             )
 
-        # Project passed in to add it to the token. Can be None.
+        send_new_code = rich.prompt.Confirm.ask(
+            "One-time authentication code required to login. Would you like to request a new code to be sent to your email?"
+        )
+
+        if send_new_code:
+            # Request 2fa email token
+            try:
+                response = requests.get(
+                    dds_cli.DDSEndpoint.REQUEST_EMAIL_2FA,
+                    auth=(self.username, password),
+                    timeout=dds_cli.DDSEndpoint.TIMEOUT,
+                )
+            except requests.exceptions.RequestException as err:
+                raise exceptions.ApiRequestError(message=str(err)) from err
+
+            LOG.info("Authentication one-time code requested, please check your email.")
+        else:
+            LOG.info("No new one-time code requested.")
+
+        one_time_password = None
+        while one_time_password is None:
+            LOG.info("Please enter the one-time authentication code (leave empty to exit):")
+            token_entered = rich.prompt.Prompt.ask("Authentication one-time code")
+            if token_entered == "":
+                raise exceptions.AuthenticationError(
+                    message="Exited due to no one-time authentication code entered."
+                )
+
+            if not token_entered.isdigit():
+                LOG.info("Please enter a valid one-time code. It should consist of only digits.")
+                continue
+            if len(token_entered) != 8:
+                LOG.info("Please enter a valid one-time code. It should consist of 8 digits.")
+                continue
+            one_time_password = token_entered
```

Comment:
I think this works really well from a users perspective, but I'm wondering if it's not possible to return the token instead of doing basic auth both times?

Reply:
I agree that the user experience of the login with mfa is carefully implemented. However, I think that a user should not be able to request 2fa, instead it is something enforced by the server. Moreover, I agree with Ina that providing user crendentials should follow with a token returned to confirm that the first factor is successfully authenticated, but we need an additional factor of authentication to follow up. For the second factor, use the token from the first factor to create the final token to be used for normal operation of the services for the given user.